### PR TITLE
V1 2 feature branch

### PR DESCRIPTION
**Pull request by emailtomanishkr**
Added support to read a key with version, added the integration test as well.

### CODE REVIEW

File: src/main/java/com/bettercloud/vault/api/Logical.java [hunk 1]
```diff
@@ -51,20 +51,32 @@ public Logical(final VaultConfig config) {
      * @throws VaultException If any errors occurs with the REST request (e.g. non-200 status code, invalid JSON payload, etc), and the maximum number of retries is exceeded.
      */
     public LogicalResponse read(final String path) throws VaultException {
+        return read(path, 0);
```
Comment (suhas0sn07) [inline src/main/java/com/bettercloud/vault/api/Logical.java]: Please specify why we are passing 0 here. Possibly with proper static final variable name

File: src/test-integration/java/com/bettercloud/vault/api/LogicalV2Tests.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        vault.logical().write(path, new HashMap<String, String>() {{
+            put(key, value_1);
+        }});
+        Assert.assertEquals(value_1, vault.logical().read(path).getData().get(key));
+        Assert.assertEquals(value_1, vault.logical().read(path, 1).getData().get(key));
+        vault.logical().write(path, new HashMap<String, String>() {{
+            put(key, value_2);
+        }});
+        Assert.assertEquals(value_2, vault.logical().read(path, 2).getData().get(key));
+        Assert.assertEquals(value_1, vault.logical().read(path, 1).getData().get(key));
```
Comment (suhas0sn07) [inline src/test-integration/java/com/bettercloud/vault/api/LogicalV2Tests.java]: Add one more assert here for default fetch, this has to return latest value.
Assert.assertEquals(value_2, vault.logical().read(path).getData().get(key));

**Comment by emailtomanishkr** [comment]
@steve-perkins  Can you please review and merge this pull request, if it looks good to you.

**Comment by steve-perkins** [comment]
Thanks for the pull request!

I do apologize for the delay in getting KV2 support added to the driver.  I've just been in crunch mode with other projects recently.  As you've noticed, I did add some primitive support to a feature branch, branched off the old `1.2` release of the library.  This was due to an immediate internal need of my employer, for a set of services that are (for the time being) pinned to that older version of the library.  

I pushed up to GitHub the source code for this internal feature branch, in case anyone in the community might care to make use of it.  However, I did NOT publish a binary artifact to Maven Central, and don't have any plans to develop this feature branch further.  

It was a stopgap measure for a very narrow use case (i.e. 100% backwards compatibility, for an older library from which we weren't ready to upgrade).  But for the mainline current version, I would like to make some different design choices.  Specifically, this feature branch performs a "pre-flight check" when a `Vault` object instantiated.  It hits a Vault endpoint and loads information about which KV version is used for every mount point.  If the authenticating principal doesn't have permission to access that Vault endpoint, then the driver absolutely fails altogether.  That was acceptable for our internal needs, but for the mainline version going forward I would like to make that behavior opt-in rather than a default.

I hope to have an update to the `3.x` generation (which will then be bumped to `4.x`) in October.  If your needs are more immediate than that, and you are comfortable being pinned to the `1.2` generation of the API for now, then my guidance would be to build your branch and publish to whichever internal artifact repository your organization is using (e.g. Artifactory, Nexus, etc).  That is essentially what we have done with our own `1.2.x` build.

**Comment by emailtomanishkr** [comment]
Hi @steve-perkins 

Thanks for the update.

At present, we're going with the approach of customizing the 1.2.x build.

The only query we've now is: Do we need to test it completely I mean the 1.2.x build or we just need to test the new changes which we're making? Since, I didn't see any unit or integration test for the Vault KV version-2.

**Comment by nathkn** [comment]
@steve-perkins has there been any movement on this yet?

**Comment by steve-perkins** [comment]
Apologies for the delay in closing the loop on this.  Support for the Vault 1.0 API, with the v2 kv engine, was added in PR #151, and is out in the current version on Maven Central.

**Pull request closed without merging**